frontend/packages/console-plugin-shared/src/components/Markdown/MarkdownView.tsx

```diff
@@ -0,0 +1,275 @@
+    allowedTags: [
+      'b',
+      'i',
+      'strike',
+      's',
+      'del',
+      'em',
+      'strong',
+      'a',
+      'p',
+      'h1',
+      'h2',
+      'h3',
+      'h4',
+      'ul',
+      'ol',
+      'li',
+      'code',
+      'pre',
+      'button',
+      'span',
+      'div',
+      ...tableTags,
+    ],
+    allowedAttributes: {
+      a: ['href', 'target', 'rel'],
+      button: ['class'],
+      i: ['class'],
+      div: ['class'],
+      span: ['class'],
+      pre: ['class'],
+      code: ['class'],
+      '*': ['data-*'],
+    },
+    allowedSchemes: ['http', 'https', 'mailto'],
+    transformTags: {
+      a: sanitizeHtml.simpleTransform('a', { rel: 'noopener noreferrer' }, true),
+    },
+  });
+};
+
+export type MarkdownProps = {
+  content?: string;
+  emptyMsg: string;
+  exactHeight?: boolean;
+  truncateContent?: boolean;
+  extensions?: ShowdownExtension[];
+  renderExtension?: (contentDocument: Document, rootSelector: string) => React.ReactNode;
+  inline?: boolean;
+  options?: ShowdownOptions;
+  theme?: string;
+  updateThemeClass?: (htmlTagElement: HTMLElement, theme: string) => void;
+};
+
+type InnerSyncMarkdownProps = Pick<
+  MarkdownProps,
+  'renderExtension' | 'exactHeight' | 'theme' | 'updateThemeClass'
+> & {
+  markup: string;
+  isEmpty: boolean;
+};
+
+export const MarkdownView: React.FC<MarkdownProps> = ({
+  truncateContent,
+  content,
+  emptyMsg,
+  extensions,
+  renderExtension,
+  exactHeight,
+  inline,
+  options,
+  theme,
+  updateThemeClass,
+}) => {
+  const markup = React.useMemo(() => {
+    const truncatedContent = truncateContent
+      ? _.truncate(content, {
+          length: 256,
+          separator: ' ',
+          omission: '\u2026',
+        })
+      : content;
+    return markdownConvert(truncatedContent || emptyMsg, extensions, options);
+  }, [content, emptyMsg, extensions, options, truncateContent]);
+  const innerProps: InnerSyncMarkdownProps = {
+    renderExtension: extensions?.length > 0 ? renderExtension : undefined,
+    exactHeight,
+    markup,
+    isEmpty: !content,
+    theme,
+    updateThemeClass,
+  };
+  return inline ? <InlineMarkdownView {...innerProps} /> : <IFrameMarkdownView {...innerProps} />;
+};
+
+type RenderExtensionProps = {
+  renderExtension: (contentDocument: Document, rootSelector: string) => React.ReactNode;
+  selector: string;
+  markup: string;
+  docContext?: Document;
+};
+
+const RenderExtension: React.FC<RenderExtensionProps> = ({
+  renderExtension,
+  selector,
+  markup,
+  docContext,
+}) => {
+  const forceRender = useForceRender();
+  const markupRef = React.useRef<string>(null);
+  const shouldRenderExtension = React.useCallback(() => {
+    if (markupRef.current === markup) {
+      return true;
+    }
+    markupRef.current = markup;
+    return false;
+  }, [markup]);
+  /**
+   * During a render cycle in which markup changes, renderExtension receives an old copy of document
```

Comment:
I know this was already existing, but if it is customer facing we should probably clean this blurb up a little.

Suggestion:

>  During a render cycle where markup changes, `renderExtension` receives an old copy of document because react is still updating the dom using `dangerouslySetInnerHTML` with latest markdown markup. Use `forceRender` to delay the rendering of the extension by one render cycle.

Reply:
Thanks! I've updated the comment